deepmd/pt/model/model/linear_model.py

```diff
@@ -0,0 +1,194 @@
+        return sorted([model.get_rcut() for model in self.models])
+
+    def get_sel(self) -> int:
+        """Get the neighbor selection."""
+        return self.get_sels()[-1]
+
+    def get_sels(self) -> int:
+        """Get the neighbor selection for each individual models in ascending order."""
+        return sorted(
+            [
+                sum(model.get_sel())
+                if isinstance(model.get_sel(), list)
+                else model.get_sel()
+                for model in self.models
+            ]
+        )
+
+    def distinguish_types(self) -> bool:
+        """If distinguish different types by sorting."""
+        return self.dp_model.distinguish_types() and self.zbl_model.distinguish_types()
+
+    def forward_atomic(
+        self,
+        extended_coord,
+        extended_atype,
+        nlist,
+        ra: float,
+        rb: float,
+        alpha: Optional[float] = 0.1,
+        mapping: Optional[torch.Tensor] = None,
+    ) -> Dict[str, torch.Tensor]:
+        """Return atomic prediction.
+
+        Note: currently only support the linear combination of a ZBL model and a DP model,
+        the weight is calculated based on this paper:
+        Appl. Phys. Lett. 114, 244101 (2019); https://doi.org/10.1063/1.5098061
+
+        Parameters
+        ----------
+        extended_coord
+            coodinates in extended region
+        extended_atype
+            atomic type in extended region
+        nlist
+            neighbor list. nf x nloc x nsel
+        mapping
+            mapps the extended indices to local indices
+        ra : float
+            inclusive lower boundary of the range in which the ZBL potential and the deep potential are interpolated.
+        rb : float
+            exclusive upper boundary of the range in which the ZBL potential and the deep potential are interpolated.
+        alpha : float
+            a tunable scale of the distances between atoms.
+
+        Returns
+        -------
+        result_dict
+            the result dict, defined by the fitting net output def.
+        """
+        # the DPAtomicModel sel is always a List or Union[List, int]?
+        nlists = build_multiple_neighbor_list(
+            extended_coord, nlist, self.get_rcuts(), self.get_sels()
+        )
+
+        zbl_nlist = nlists[str(self.zbl_model.rcut) + "_" + str(self.zbl_model.sel)]
+        dp_nlist = nlists[
+            str(self.dp_model.rcut) + "_" + str(self.dp_model.sel)
+        ]  # need to handle sel dtype.
+
+        zbl_nframe, zbl_nloc, zbl_nnei = zbl_nlist.shape
+        dp_nframe, dp_nloc, dp_nnei = dp_nlist.shape
+        zbl_atype = extended_atype[
+            :, :zbl_nloc
+        ]  # nframe, nloc should all be the same, only difference is nnei based on rcut and nlist.
+        dp_atype = extended_atype[:, :dp_nloc]
+
+        # which rr should I use? this rr should be (nfrmaes, nloc, nnei)
+        zbl_weight = self._compute_weight(rr, ra, rb)
```

Comment:
@wanghan-iapcm If I understand this correctly, I think `nframe` and `nloc` should be the same for two sub atomic models, `nnei` will be different based on their respective `rcut`s and `nlist`s. When calculating the zbl weight, we need to use `rr`, which has the same shape as `nlist`. which nlist should we use?

Reply:
Yes correct. we prefer the larger rcut to calculate the conf-determined weights.